### ISSUE TITLE
Small Optimizations in general

### ISSUE THREAD

**Issue by Speiger** [seinecle/MapsOfScience#1]
https://github.com/seinecle/MapsOfScience/blob/c9b979c05e6472fd8f53f9e6b1ef11afa9e620f8/src/main/java/net/clementlevallois/functions/mapsofscience/JournalSimilaritiesComputer.java#L117

This could be a computeIfAbsent which requests a key and a Function that creates a "defaultValue" if it isn't present using a given key.
(Or if you are using PrimitiveCollections its SupplyIfAbsent which just provides a defaultValue ignoring the key, which is if the defaultValue never changes also useful)

https://github.com/seinecle/MapsOfScience/blob/c9b979c05e6472fd8f53f9e6b1ef11afa9e620f8/src/main/java/net/clementlevallois/functions/mapsofscience/JournalSimilaritiesComputer.java#L27

This can be a Long2ObjectMap<LongSet>. Yes in your article you say the "boxing/unboxing" is inconvenient and may cause other issues.
But in your code for example you NEVER have a case where that actually would apply.
There is functions like: nextLong which just returns the primitiveType without boxing/unboxing it at all.

Another thing is:
Instead of "OpenHashSet" the "LinkedOpenHashSet" would be a performance improvement too since you do a LOT of iteration.
And LinkedHashSets have a LOT of a speedups  since they just can ignore gaps in the Bucket.

One thing is: I wonder what this function would do to your performance:
https://github.com/Speiger/Primitive-Collections/blob/debug/src/main/java/speiger/src/collections/longs/collections/LongIterable.java#L335

It would turn your code into this:
```java
private int computeSimilarities(LongSet first, LongSet second) {
   return first.size() < second.size() ? first.count(second::contains) : second.count(first::contains);
}
```

One last thing. Streams while no longer as expensive as in java8, should be avoided still to some degree.
They create a backing state machine, which creates a lot of extra overhead that you may not need.
Just add a breakpoint inside the for each. It can get gigantic real quick.

Oh yeah  also the skip is not needed you could just start at index a +1 out of the gate. No need to iterate empty.

**Comment by seinecle** [comment]
Thanks a lot!

Implemented some of your suggestions [here](https://github.com/seinecle/MapsOfScience/blob/24eeac4f8eb8eca443b15c5bd5d3ee17c7089e5b/src/main/java/net/clementlevallois/functions/mapsofscience/JournalSimilaritiesComputer.java).

did no remove streams as parallelstreams provide speedup.
did not try your collection because... lack of time to switch from fastutil.

These changes brought a 4% improvement:

![updated table](https://github.com/seinecle/MapsOfScience/assets/1244100/1b86662d-23ab-4d93-a327-e8eff2abb98e)

**Comment by Speiger** [comment]
No Problem.

that was more of a performance improvement then i expected.
For that little change.

Yeah I didn't expect you to do that. Since that is a big ask in the first place.
Though transition between fastutil/primitive collections is in 95% of the cases just a "Reimport" classnames/functions are the same if present on both sides.

I wonder how Primitive iterators & Primitive Only sets would suffice here.
But this is no rush and or requirement, just curiosity.

One thing I would highly suggest though:
Unless you need a Direct Implementation feature, please abstain from using "Hardcoded implementations".
In other words: You want to compare 2 Sets, Do you actually need: "OpenLinkedHashSet" as parameter? Or is "ObjectSet" not enough? Or Set?

Because it would also reduce Changes required if you do a drastic change like you did here.
All these extra changes wouldn't be necessary otherwise if only Set/ObjectSet were used in the first place instead of ObjectOpenHashSet.

To be fair its a mentality, but not a bad one IMO.

**Comment by seinecle** [comment]
You are right this is definitely a good practice, will update.

**Comment by seinecle** [comment]
Closing the issue, don't hesitate opening a new one if necessary.
